modules/core/src/main/scala/com.snowplowanalytics.snowplow/scalatracker/Tracker.scala

```diff
@@ -83,23 +75,24 @@ class Tracker(emitters: Seq[TEmitter], namespace: String, appId: String, encodeB
                               contexts: Seq[SelfDescribingJson],
                               timestamp: Option[Timestamp]): Payload = {
 
-    payload.add("eid", UUID.randomUUID().toString)
+    val newPayload = payload
+      .add("eid", UUID.randomUUID().toString)
+      .add("tv", Version)
+      .add("tna", namespace)
+      .add("aid", appId)
+      .addDict(subject.subjectInformation)
 
-    if (!payload.nvPairs.contains("dtm")) {
+    val payloadWithTimestamp = if (!newPayload.get.contains("dtm")) {
       timestamp match {
-        case Some(DeviceCreatedTimestamp(dtm)) => payload.add("dtm", dtm.toString)
-        case Some(TrueTimestamp(ttm))          => payload.add("ttm", ttm.toString)
-        case None                              => payload.add("dtm", System.currentTimeMillis().toString)
+        case Some(DeviceCreatedTimestamp(dtm)) => newPayload.add("dtm", dtm.toString)
+        case Some(TrueTimestamp(ttm))          => newPayload.add("ttm", ttm.toString)
+        case None                              => newPayload.add("dtm", System.currentTimeMillis().toString)
```

Comment:
`System.currentTimeMillis()` is a problem. It breaks the RT. 

And this "time-updating" is an interesting problem that I struggled a lot in Haskell tracker (and ended up making everything `IO`-based). It should be easy to strip it away with some kind of `Clock`-algebra, that by default is interpreted into `IO` (upcoming version of cats-effect [will include that](https://github.com/typelevel/cats-effect/pull/289/commits/dfc80ec2eecc2bae7f2b4c419d3ed2221fe189f9), btw).

But right now `completePayload` should have `F[_]: Applicative` signature, which will be propagated to all `track`-methods.

Reply:
To illustrate the problem, here's a small snippet:

```scala
  def completePayload(payload: Map[String, String]) =
    payload ++ Map("time" -> System.currentTimeMillis().toString)

  def trackBroken: IO[Unit] =
    send(completePayload(Map.empty))

  def trackPure: IO[Unit] =
    IO(completePayload(Map.empty)).flatMap(send)

  def send(payload: Map[String, String]): IO[Unit] =
    IO(Thread.sleep(100)) *> IO(println(s"Sending $payload"))

  List.fill(10)(trackBroken).sequence.void.unsafeRunSync()  // prints same timestamp - RT is broken
  List.fill(10)(trackPure).sequence.void.unsafeRunSync()   // prints timestamps according to delay 
```

Mixing eager and lazy evaluations always lead to worst kind of bugs.

Btw, missed in the diff - same problem applies to `UUID` above.